src/main/java/org/springframework/integration/kafka/outbound/KafkaProducerMessageHandler.java

```diff
@@ -136,53 +183,104 @@ public void setSendTimeoutExpression(Expression sendTimeoutExpression) {
+				this.partitionIdExpression.getValue(this.evaluationContext, requestMessage, Integer.class)
+				: requestMessage.getHeaders().get(KafkaHeaders.PARTITION_ID, Integer.class);
 
 		Object messageKey = this.messageKeyExpression != null
-				? this.messageKeyExpression.getValue(this.evaluationContext, message)
-				: message.getHeaders().get(KafkaHeaders.MESSAGE_KEY);
+				? this.messageKeyExpression.getValue(this.evaluationContext, requestMessage)
+				: requestMessage.getHeaders().get(KafkaHeaders.MESSAGE_KEY);
+
+		if (this.replyContainer != null) {
+			if (messageKey != null) {
+				if (logger.isDebugEnabled()) {
+					logger.debug("Key '" + messageKey + "' ignored for request/reply");
+				}
+			}
+			messageKey = UUID.randomUUID().toString();
+			this.pendingReplies.put((String) messageKey, new SettableListenableFuture<>());
+		}
 
 		Long timestamp = this.timestampExpression != null
-				? this.timestampExpression.getValue(this.evaluationContext, message, Long.class)
-				: message.getHeaders().get(KafkaHeaders.TIMESTAMP, Long.class);
+				? this.timestampExpression.getValue(this.evaluationContext, requestMessage, Long.class)
+				: requestMessage.getHeaders().get(KafkaHeaders.TIMESTAMP, Long.class);
 
-		V payload = (V) message.getPayload();
+		V payload = (V) requestMessage.getPayload();
 		if (payload instanceof KafkaNull) {
 			payload = null;
 		}
 
 		ListenableFuture<?> future = this.kafkaTemplate.send(topic, partitionId, timestamp, (K) messageKey, payload);
 
 		if (this.sync) {
-			Long sendTimeout = this.sendTimeoutExpression.getValue(this.evaluationContext, message, Long.class);
+			waitForAck(requestMessage, future);
+		}
+		if (this.replyContainer != null) {
+			return obtainReply(requestMessage, messageKey);
+		}
+		else {
+			return null;
+		}
+	}
+
+	private void waitForAck(final Message<?> requestMessage, ListenableFuture<?> future) {
+		Long sendTimeout = this.sendTimeoutExpression.getValue(this.evaluationContext, requestMessage, Long.class);
+		try {
 			if (sendTimeout == null || sendTimeout < 0) {
 				future.get();
 			}
 			else {
-				try {
-					future.get(sendTimeout, TimeUnit.MILLISECONDS);
-				}
-				catch (TimeoutException te) {
-					throw new MessageTimeoutException(message, "Timeout waiting for response from KafkaProducer", te);
-				}
+				future.get(sendTimeout, TimeUnit.MILLISECONDS);
 			}
 		}
+		catch (TimeoutException te) {
+			throw new MessageTimeoutException(requestMessage, "Timeout waiting for response from KafkaProducer", te);
+		}
+		catch (InterruptedException e) {
+			Thread.currentThread().interrupt();
+			throw new MessageHandlingException(requestMessage, e);
+		}
+		catch (ExecutionException e) {
+			throw new MessageHandlingException(requestMessage, e);
+		}
+	}
+
+	private Object obtainReply(final Message<?> requestMessage, Object messageKey) {
+		try {
+			ConsumerRecord<K, V> reply = this.pendingReplies.get(messageKey).get(this.receiveTimeout,
```

Comment:
I wonder if `AbstractMessageProducingHandler.async` can be used here.
At least we already have that `sync` option which may dictate us what to do with this `ListenableFuture`
WDYT?